src/util/listQueryBuilder.js

```diff
@@ -451,6 +447,119 @@ async function performAggregationQuery ({
+ * @param {Object}  [params.filters[key]]
+ * @param {String}  [params.filters[key].dbField]
+ * @param {String}  [params.filters[key].type]
+ * @param {String|Function}  [params.filters[key].transformValue]
+ * @param {String|Function}  [params.filters[key].query]
+ *
+ * @param {Object}  params.paginationConfig
+ * @param {Number}  params.paginationConfig.page
+ * @param {Number}  params.paginationConfig.nbResultsPerPage
+ *
+ * @param {Object}  params.orderConfig
+ * @param {Object}  params.orderConfig.order - Must be 'asc' or 'desc'
+ *
+ * @return {Object} paginationMeta
+ * @return {Number} paginationMeta.nbResults
+ * @return {Number} paginationMeta.nbPages
+ * @return {Number} paginationMeta.page
+ * @return {Number} paginationMeta.nbResultsPerPage
+ * @return {Object[]} paginationMeta.results
+ */
+async function performHistoryQuery ({
+  queryBuilder,
+  schema,
+  table,
+  timeFilter = 'createdDate',
+  timeColumn = 'createdTimestamp',
+  retentionLimitDate,
+  groupBy,
+  filters,
+  paginationConfig,
+  orderConfig
+}) {
+  checkOrderConfig(orderConfig)
+
+  let onlyTimeFilter = true
+
+  if (filters) {
+    checkFilters(filters)
+
+    const nonTimeFilterKeys = _.without(Object.keys(filters), [timeFilter])
+    onlyTimeFilter = nonTimeFilterKeys.map(key => _.get(filters, `${key}.value`)).every(_.isUndefined)
+    const minRangeValue = getMinRangeValue(filters[timeFilter].value)
+
+    if (minRangeValue && minRangeValue < retentionLimitDate && !onlyTimeFilter) {
+      throw createError(400, `All filters are disabled before ${retentionLimitDate}`)
+    }
+  }
+  checkPaginationConfig(paginationConfig)
+
+  if (!['hour', 'day', 'month'].includes(groupBy)) throw new Error(`Invalid groupBy value: ${groupBy}`)
+
+  const intervals = {
+    hour: '1 hour',
+    day: '1 day',
+
+    // TimescaleDB doesn't support month interval (https://github.com/timescale/timescaledb/issues/414)
+    month: '30 days'
+  }
+
+  const knex = queryBuilder
+
+  const interval = intervals[groupBy]
+
+  // `AT TIME ZONE 'UTC'` statement is important to add to have the type 'timestamptz'
+  // otherwise the type 'timestamp' is returned and the date won't be UTC
+  const select = knex.raw(`public.time_bucket(INTERVAL '${interval}', ??) AT TIME ZONE 'UTC' as ??`, [timeColumn, groupBy])
+
+  queryBuilder = queryBuilder
+    .select(select)
+    .from(knex.raw('??.??', [schema, table]))
+    .count()
+    .groupBy(groupBy)
+
+  let transformedValues
+
+  if (filters) {
+    transformedValues = getTransformedValues(filters)
+    addFiltersToQueryBuilder(queryBuilder, filters, transformedValues)
+  }
+
+  const { orderBy, order } = orderConfig
+  queryBuilder.orderBy(orderBy, order)
+
+  const { page, nbResultsPerPage } = paginationConfig
+
+  const allResults = await queryBuilder
+
+  // To avoid query complexity, data retrieval for pagination aren't split into results query and count query.
+  // TODO: improve it with the incoming cursor-based pagination
```

Comment:
I see you’re using `count()` above, which would mean that you get both results and count according to your comment?
Why not do the same for other models?

Reply:
Actually this `count()` is the data we want to retrieve in rows, not the pagination count.
To have the pagination count, we need to do a `count` of `count`, which I chose to not do because of incoming cursor pagination which don't perform this count request.